zkml/src/layers/transformer/softmax.rs

```diff
@@ -1,60 +1,361 @@
+        // 1) dequantise the input
+        // 2) apply a float mask
+        // 3) sum along the desired dim
+        let padding_mode = if input.shape != unpadded_input_shape {
+            PaddingMode::Padding
+        } else {
+            PaddingMode::NoPadding
+        };
+
+        let float_mask =
+            AttentionMask::<f32>::new(unpadded_input_shape, f32::NEG_INFINITY, padding_mode)?;
+
+        let dequantised_input = input.dequantize(input_scale_factor);
+        let masked_input = float_mask.apply(&dequantised_input)?;
+
+        let chunk_size = *masked_input
+            .shape
+            .last()
+            .ok_or(anyhow!("Input tensor had no shape in quantised Softmax"))?;
+
+        // Now that we ahve the masked input in float calculate the shift
+        let shift_data = masked_input
+            .get_data()
+            .chunks(chunk_size)
+            .map(|vec| {
+                let sum = vec
+                    .iter()
+                    .map(|x| (x / float_temperature).exp())
+                    .sum::<f32>();
+                let log_sum = sum.ln();
+                let shift = -(SCALE_FACTOR as f32 * float_temperature * log_sum).round() as Element;
+                vec![shift; chunk_size]
+            })
+            .flatten()
+            .collect::<Vec<_>>();
+
+        let shift_tensor = Tensor::<Element>::new(input.get_shape(), shift_data);
+
+        let negative_infinity = -((bkm >> 16) + 1) << 16;
+        let AttentionMask { tril, bias, .. } =
+            AttentionMask::<Element>::new(unpadded_input_shape, negative_infinity, padding_mode)?;
+
+        let masked_shift = shift_tensor.mul(&tril).add(&bias);
+        // This is annoying but we now have to go over the shifted input, we need all of the entries after applying the shift to be negative
+        // and due to rounding on rows where we only care about the first input (so the first token in the sequence) it can happen that
+        // rounding means we would get a positive value here. To improve accuracy we correct for this so that these values should always get mapped to 1.
+        let mut shift_mask = AttentionMask {
+            tril: tril.scalar_mul(&self.scalar),
+            bias: masked_shift,
+            padding_mode,
+            negative_infinity,
+        };
+
+        let shifted_input = shift_mask.apply(input)?;
+        // Find all the entires that are positive (this will only be entries on rows where we normalise over 1 element)
+        // if they are positive we edit `shift_mask.bias` to correct for it.
+        shift_mask
+            .bias
+            .data
+            .iter_mut()
+            .zip(shifted_input.get_data().iter())
+            .for_each(|(bias_v, v)| {
+                if v.is_positive() {
+                    *bias_v -= v;
+                }
+            });
+
+        // We also multiply the tril tensor by `self.scalar` so that we can just call `mask.apply` on the input
+        Ok(shift_mask)
```

Comment:
related to previous comment but can't we simply put that in `apply()` instead of change the inner data of the mask ?

Reply:
Do you mean you will change the code ?